src/NuGet.Core/NuGet.Commands/RestoreCommand/LockFileBuilder.cs

```diff
@@ -551,5 +550,49 @@ private static bool HasTools(string file)
         {
             return file.StartsWith("tools/", StringComparison.OrdinalIgnoreCase);
         }
+
+        /// <summary>
+        /// An <see cref="IEqualityComparer{T}" /> that compares <see cref="LockFileTargetLibrary" /> objects by the value of the <see cref="LockFileTargetLibrary.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties.
+        /// </summary>
+        private class LockFileTargetLibraryNameAndVersionEqualityComparer : IEqualityComparer<LockFileTargetLibrary>
+        {
+            /// <summary>
+            /// Gets a static singleton for the <see cref="LockFileTargetLibraryNameAndVersionEqualityComparer" /> class.
+            /// </summary>
+            public static LockFileTargetLibraryNameAndVersionEqualityComparer Instance = new();
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="LockFileTargetLibraryNameAndVersionEqualityComparer" /> class.
+            /// </summary>
+            private LockFileTargetLibraryNameAndVersionEqualityComparer()
+            {
+            }
+
+            /// <summary>
+            /// Determines whether the specified <see cref="LockFileTargetLibrary" /> objects are equal by comparing their <see cref="LockFileTargetLibrary.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties.
+            /// </summary>
+            /// <param name="x">The first <see cref="LockFileTargetLibrary" /> to compare.</param>
+            /// <param name="y">The second <see cref="LockFileTargetLibrary" /> to compare.</param>
+            /// <returns><c>true</c> if the specified <see cref="LockFileTargetLibrary" /> objects' <see cref="FileSystemInfo.Name" /> and <see cref="LockFileTargetLibrary.Version" /> properties are equal, otherwise <c>false</c>.</returns>
```

Comment:
Missed a reference to FileSystemInfo after copy/paste 🙃

Reply:
Fixed in the latest commit.